src/host/_stream.cpp

```diff
@@ -126,565 +165,286 @@ void AdjustCursorPosition(SCREEN_INFORMATION& screenInfo, _In_ til::point coordC
+                // that was previously padded with a whitespace in the last column due to a wide glyph.
+                // Now that the wide glyph is presumably gone, we can move up a line.
+                if (pos.x == 0 && pos.y != 0 && textBuffer.GetRowByOffset(pos.y - 1).WasDoubleBytePadded())
+                {
+                    moveUp();
                 }
                 else
                 {
-                    CursorPosition.x--;
+                    AdjustCursorPosition(screenInfo, pos, keepCursorVisible, psScrollY);
                 }
-            }
-            if ((dwFlags & WC_LIMIT_BACKSPACE) && (CursorPosition.x < 0))
-            {
-                CursorPosition.x = 0;
-                OutputDebugStringA(("CONSRV: Ignoring backspace to previous line\n"));
-            }
-            AdjustCursorPosition(screenInfo, CursorPosition, (dwFlags & WC_KEEP_CURSOR_VISIBLE) != 0, psScrollY);
-            if (dwFlags & WC_DESTRUCTIVE_BACKSPACE)
-            {
-                try
+
+                // Notify accessibility to read the backspaced character.
+                // See GH:12735, MSFT:31748387
+                if (screenInfo.HasAccessibilityEventing())
                 {
-                    screenInfo.Write(OutputCellIterator(UNICODE_SPACE, Attributes, 1), cursor.GetPosition());
+                    if (const auto pConsoleWindow = ServiceLocator::LocateConsoleWindow())
+                    {
+                        LOG_IF_FAILED(pConsoleWindow->SignalUia(UIA_Text_TextChangedEventId));
+                    }
                 }
-                CATCH_LOG();
+                continue;
             }
-            if (cursor.GetPosition().x == 0 && fWrapAtEOL && pwchBuffer > pwchBufferBackupLimit)
+            case UNICODE_TAB:
             {
-                if (CheckBisectProcessW(screenInfo,
-                                        pwchBufferBackupLimit,
-                                        pwchBuffer + 1 - pwchBufferBackupLimit,
-                                        gsl::narrow_cast<size_t>(coordScreenBufferSize.width) - sOriginalXPosition,
-                                        sOriginalXPosition,
-                                        dwFlags & WC_PRINTABLE_CONTROL_CHARS))
-                {
-                    CursorPosition.x = coordScreenBufferSize.width - 1;
-                    CursorPosition.y = cursor.GetPosition().y - 1;
-
-                    // since you just backspaced yourself back up into the previous row, unset the wrap flag
-                    // on the prev row if it was set
-                    textBuffer.GetRowByOffset(CursorPosition.y).SetWrapForced(false);
-
-                    AdjustCursorPosition(screenInfo, CursorPosition, dwFlags & WC_KEEP_CURSOR_VISIBLE, psScrollY);
-                }
+                const auto pos = cursor.GetPosition();
+                const auto tabCount = gsl::narrow_cast<size_t>(NUMBER_OF_SPACES_IN_TAB(pos.x));

```

Comment:
Is `NUMBER_OF_SPACES_IN_TAB` used anywhere else? If not, should we just move it into this file to improve mental locality?

Reply:
Yeah it's used in a number of other places as well. I'd keep it for a bit, because it's a fun historical artifact. 😅